storage/innobase/handler/ha_innodb.cc

```diff
@@ -9023,91 +9022,100 @@ ha_innobase::index_read(
 			m_prebuilt->srch_key_val_len,
 			index,
 			(byte*) key_ptr,
-			(ulint) key_len);
+			key_len);
 
 		DBUG_ASSERT(m_prebuilt->search_tuple->n_fields > 0);
 	} else {
+		ut_ad(find_flag != HA_READ_KEY_EXACT);
 		/* We position the cursor to the last or the first entry
 		in the index */
 
 		dtuple_set_n_fields(m_prebuilt->search_tuple, 0);
 	}
 
-	page_cur_mode_t	mode = convert_search_mode_to_innobase(find_flag);
-
-	ulint	match_mode = 0;
-
-	if (find_flag == HA_READ_KEY_EXACT) {
-
-		match_mode = ROW_SEL_EXACT;
-
-	} else if (find_flag == HA_READ_PREFIX_LAST) {
+	page_cur_mode_t	mode = PAGE_CUR_LE;
+	m_last_match_mode = 0;
 
-		match_mode = ROW_SEL_EXACT_PREFIX;
+	switch (find_flag) {
+	case HA_READ_KEY_EXACT:
+		/* this does not require the index to be UNIQUE */
+		m_last_match_mode = ROW_SEL_EXACT;
+		/* fall through */
+	case HA_READ_KEY_OR_NEXT:
+		mode = PAGE_CUR_GE;
+		break;
+	case HA_READ_AFTER_KEY:
+		mode = PAGE_CUR_G;
+		break;
+	case HA_READ_BEFORE_KEY:
+		mode = PAGE_CUR_L;
+		break;
+	case HA_READ_PREFIX_LAST:
+		m_last_match_mode = ROW_SEL_EXACT_PREFIX;
+		break;
+	case HA_READ_KEY_OR_PREV:
+	case HA_READ_PREFIX_LAST_OR_PREV:
+		break;
+	case HA_READ_MBR_CONTAIN:
+		mode = PAGE_CUR_CONTAIN;
+		break;
+	case HA_READ_MBR_INTERSECT:
+		mode = PAGE_CUR_INTERSECT;
+		break;
+	case HA_READ_MBR_WITHIN:
+		mode = PAGE_CUR_WITHIN;
+		break;
+	case HA_READ_MBR_DISJOINT:
+		mode = PAGE_CUR_DISJOINT;
+		break;
+	case HA_READ_MBR_EQUAL:
+		mode = PAGE_CUR_MBR_EQUAL;
+		break;
+	case HA_READ_PREFIX:
+		table->status = STATUS_NOT_FOUND;
+		DBUG_RETURN(HA_ERR_UNSUPPORTED);
 	}
```

Comment:
After comparing with  convert_search_mode_to_innobase() I find one difference i.e. when the  flag doesn't match with any of the cases.
```
page_cur_mode_t
convert_search_mode_to_innobase(
/*============================*/
        ha_rkey_function        find_flag)
    ...
        my_error(ER_CHECK_NOT_IMPLEMENTED, MYF(0), "this functionality");
        return(PAGE_CUR_UNSUPP);
```
Can we add it as the default case here to ensure zero functional impact ?

Reply:
GCC at least since version 3 implements a warning if there is a `switch` on an `enum` variable and some constant is not handled. I tested it by removing one of the `case` labels:
```
/mariadb/10.6/storage/innobase/handler/ha_innodb.cc:9040:16: warning: enumeration value ‘HA_READ_KEY_OR_PREV’ not handled in switch [-Wswitch]
```
This useful compile-time check would be disabled if we added a `default:` label. I think that it is better to list all `enum` values so that whenever someone decides to add one more `enum` values, there will be a warning that will force this code to be adjusted.

Reply:
I am going to retain `convert_search_mode_to_innobase()`, and the `return true` (to indicate the unsupported code path) will be outside the `switch` statement.